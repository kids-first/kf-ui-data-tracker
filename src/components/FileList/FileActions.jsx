import React from 'react';
import PropTypes from 'prop-types';
import classes from 'classnames';
import {Mutation} from 'react-apollo';
import {GET_STUDY_BY_ID} from '../../state/queries';
import {DELETE_FILE, FILE_DOWNLOAD_URL} from '../../state/mutations';
import {KF_STUDY_API} from '../../common/globals';
import {Icon} from 'kf-uikit';

const FileActions = ({node, studyId, className}) => {
  const FileActionsClass = classes('FileActions', className);

  const FileActionButton = classes('FileActionButton');
  if (node) {
    return (
      <Mutation mutation={FILE_DOWNLOAD_URL} key={node.kfId}>
        {downloadFile => (
          <Mutation
            mutation={DELETE_FILE}
            update={(cache, {data: {deleteFile}}) => {
              // Re-writes the study in the cache with the deleted file
              // removed.
              const {studyByKfId} = cache.readQuery({
                query: GET_STUDY_BY_ID,
                variables: {kfId: studyId},
              });
              const deleteIndex = studyByKfId.files.edges.findIndex(
                edge => edge.node.kfId === deleteFile.kfId,
              );
              if (deleteIndex < 0) {
                return studyByKfId;
              }
              studyByKfId.files.edges.splice(deleteIndex, 1);
              const data = {
                studyByKfId: {
                  ...studyByKfId,
                  files: {
                    __typename: 'FileNodeConnection',
                    edges: studyByKfId.files.edges,
                  },
                },
              };
              cache.writeQuery({
                query: GET_STUDY_BY_ID,
                data,
              });
            }}
            key={node.kfId}
          >
            {(deleteFile, {loading, error}) => (
              <FileActionButtons
                {...{
                  node,
                  studyId,
                  deleteFile,
                  downloadFile,
                  loading,
                  className,
                  error,
                }}
              />
            )}
          </Mutation>
        )}
      </Mutation>
    );
  }
  return null;
};

export default FileActions;
