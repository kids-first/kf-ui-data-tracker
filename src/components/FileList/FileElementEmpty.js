import React from 'react';
import PropTypes from 'prop-types';
import classes from 'classnames';
import Badge from '../Badge/Badge';
/**
 * Displays unordered studies in grid view (include empty stage message)
 */

const FileElementEmpty = ({className}) => {
  const fileElementEmptyClass = classes(
    'FileList--Element',
    'sm:flex-no-wrap',
    className,
  );

  return (
    <li className={fileElementEmptyClass}>
      <div className="FileList--ElementBadge sm:w-48">
        <Badge state="empty" />
        <Badge state="empty" />
      </div>
      <div className="flex-initial">
        <p className="mt-0 pt-2 font-bold text-sm">
          <span className="bg-lightGrey px-12 mx-4" />
          <span className="bg-lightGrey px-8 mx-4" />
          <span className="bg-lightGrey px-12 mx-4" />
        </p>
        <p className="mt-0 font-normal text-grey text-xs">
          <span className="bg-lightGrey px-12 mx-4" />
          <span className="bg-lightGrey px-8 mx-4" />
          <span className="bg-lightGrey px-6 mx-4" />
        </p>
      </div>
    </li>
  );
};

FileElementEmpty.propTypes = {
  /** Any additional classes to be applied to the empty study element*/
  className: PropTypes.string,
};

FileElementEmpty.defaultProps = {
  className: null,
};

export default FileElementEmpty;
