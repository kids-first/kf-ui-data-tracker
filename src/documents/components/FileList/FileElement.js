import React, {useState, useEffect} from 'react';
import PropTypes from 'prop-types';
import {withRouter, Link} from 'react-router-dom';
import TimeAgo from 'react-timeago';
import {Table, Icon, Checkbox, Popup} from 'semantic-ui-react';
import Markdown from 'react-markdown';
import FileActionButtons from '../FileActionButtons/FileActionButtons';
import {fileSortedVersions, fileLatestDate} from '../../utilities';
import {fileTypeDetail} from '../../../common/enums';
import {longDate} from '../../../common/dateUtils';
import FileTags from '../FileDetail/FileTags';

const useRecentlyUpdated = (latestDate, fileId) => {
  const [justUpdated, setJustUpdated] = useState(false);

  useEffect(() => {
    let FIVE_MIN = 300000;
    let docs = sessionStorage.getItem('kf_updated_docs');
    // make sure we only select one row
    if (fileId === docs) {
      // make sure that row was updated less than 5 minutes ago
      if (new Date().getTime() - Date.parse(latestDate) < FIVE_MIN) {
        setJustUpdated(true);
        setTimeout(() => {
          // reset after 7seconds
          setJustUpdated(false);
          sessionStorage.removeItem('kf_updated_docs');
        }, 7000);
      }
    }
    return () => {
      setJustUpdated(false);
    };
  }, [latestDate, fileId]);

  return justUpdated;
};

/**
 * Displays unordered study files in list view
 */
const FileElement = ({
  fileNode,
  loading,
  history,
  match,
  fileListId,
  updateFile,
  deleteFile,
  downloadFileMutation,
  selected,
  onSelectOne,
}) => {
  const fileKfID = fileNode.kfId || 'unknown ID';
  const fileName = fileNode.name || 'unknown file name';
  // Show at most 5 lines of the description
  const fileDescription =
    fileNode.description
      .split('\n')
      .splice(0, 5)
      .join('\n') || '';
  const sortedVersions = fileSortedVersions(fileNode);
  const latestDate = fileLatestDate(sortedVersions);
  const fileType =
    fileNode && fileTypeDetail[fileNode.fileType]
      ? fileTypeDetail[fileNode.fileType]
      : {title: 'unknown', icon: 'question'};
  const justUpdated = useRecentlyUpdated(latestDate, fileKfID);
  return (
    <Table.Row
      style={{backgroundColor: justUpdated ? '#f8ffff' : 'inherit'}}
      data-testid="file-item"
      className="cursor-pointer"
      onClick={() =>
        history.push(`/study/${match.params.kfId}/documents/${fileKfID}`)
      }
    >
      <Table.Cell
        textAlign="center"
        onClick={e => {
          e.stopPropagation();
          onSelectOne(fileKfID);
        }}
      >
        <Checkbox data-testid="file-select" checked={selected} />
      </Table.Cell>
      <Popup
        wide="very"
        mouseEnterDelay={500}
        header={fileType.title || 'Unknown Type'}
        content={fileType.description || 'Unknown Type'}
        trigger={
          <Table.Cell textAlign="center">
            <Icon name={fileType.icon || 'question'} />
          </Table.Cell>
        }
      />
      <Table.Cell>
        {fileDescription.length > 3 && (
          <Popup
            offset="-75%"
            wide="very"
            position="top left"
            trigger={<Icon name="info circle" />}
            content={
              <Markdown
                source={fileDescription}
                renderers={{
                  image: Image,
                  table: props => <Table>{props.children}</Table>,
                }}
              />
            }
          />
        )}
        <Link to={`/study/${match.params.kfId}/documents/${fileKfID}`}>
          {fileName}
        </Link>
      </Table.Cell>
      <Table.Cell textAlign="center" width="4">
        <FileTags fileNode={fileNode} updateFile={updateFile} />
      </Table.Cell>
      <Table.Cell textAlign="center" width="1">
        <TimeAgo date={latestDate} live={false} title={longDate(latestDate)} />
      </Table.Cell>
      <Table.Cell textAlign="center">
        <FileActionButtons
          fluid
          node={fileNode}
          studyId={fileListId}
          deleteFile={deleteFile}
          downloadFileMutation={downloadFileMutation}
        />
      </Table.Cell>
    </Table.Row>
  );
};

FileList.propTypes = {
  /** Array of study object*/
  fileNode: PropTypes.object,
};

FileList.defaultProps = {
  fileNode: null,
};

export default withRouter(FileElement);
