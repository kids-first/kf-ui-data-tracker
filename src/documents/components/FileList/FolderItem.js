import {Button, Icon, List} from 'semantic-ui-react';
import {Link, withRouter} from 'react-router-dom';
import React, {useState} from 'react';

import TimeAgo from 'react-timeago';
import {longDate} from '../../../common/dateUtils';

/**
 * Displays study files in folder view
 */

const FolderItem = ({
  nodeProps,
  searchString,
  setSearchString,
  updateFile,
  deleteFile,
  studyId,
  setInputOpen,
  setRenameOpen,
  setDeleteOpen,
}) => {
  const {
    scaffoldBlockPxWidth,
    toggleChildrenVisibility,
    connectDragPreview,
    connectDragSource,
    isDragging,
    canDrop,
    canDrag,
    node,
    draggedNode,
    path,
    treeIndex,
    isSearchMatch,
    isSearchFocus,
    className,
    style,
    didDrop,
  } = nodeProps;

  const [hover, setHover] = useState(false);

  let handle;
  if (canDrag) {
    if (typeof node.children === 'function' && node.expanded) {
      // Show a loading symbol on the handle when the children are expanded
      //  and yet still defined by a function (a callback to fetch the children)
      handle = (
        <div className="rst__loadingHandle">
          <div className="rst__loadingCircle">
            {[...new Array(12)].map((_, index) => (
              <div
                // eslint-disable-next-line react/no-array-index-key
                key={index}
                className="rst__loadingCirclePoint"
              />
            ))}
          </div>
        </div>
      );
    } else {
      // Show the handle used to initiate a drag-and-drop
      handle = connectDragSource(
        <div className="rst__moveHandle">
          {node.isDirectory ? (
            <Icon
              className="px-10 py-9"
              size="large"
              name={node.expanded ? 'folder open' : 'folder'}
              color={isSearchMatch ? 'blue' : 'black'}
            />
          ) : (
            <Icon
              className="px-10 pt-5"
              size="large"
              name="file alternate"
              color={isSearchMatch ? 'blue' : 'black'}
            />
          )}
        </div>,
        {
          dropEffect: 'copy',
        },
      );
    }
  }
  const isDescendant = (older, younger) => {
    return (
      !!older.children &&
      typeof older.children !== 'function' &&
      older.children.some(
        child => child === younger || isDescendant(child, younger),
      )
    );
  };
  const isDraggedDescendant = draggedNode && isDescendant(draggedNode, node);
  const isLandingPadActive = !didDrop && isDragging;

  let buttonStyle = {left: -0.5 * scaffoldBlockPxWidth};
  return (
    <div style={{height: '100%'}}>
      {toggleChildrenVisibility &&
        node.children &&
        (node.children.length > 0 || typeof node.children === 'function') && (
          <div>
            <button
              type="button"
              aria-label={node.expanded ? 'Collapse' : 'Expand'}
              className={
                node.expanded ? 'rst__collapseButton' : 'rst__expandButton'
              }
              style={buttonStyle}
              onClick={() => {
                toggleChildrenVisibility({
                  node,
                  path,
                  treeIndex,
                });
                setSearchString('');
              }}
            />
            {node.expanded && !isDragging && (
              <div
                style={{width: scaffoldBlockPxWidth}}
                className={'rst__lineChildren'}
              />
            )}
          </div>
        )}

      <div className={'rst__rowWrapper'}>
        {/* Set the row preview to be used during drag and drop */}
        {connectDragPreview(
          <div
            className={[
              'rst__row',
              isLandingPadActive && 'rst__rowLandingPad',
              isLandingPadActive && !canDrop && 'rst__rowCancelPad',
              isSearchMatch && 'rst__rowSearchMatch',
              isSearchFocus && 'rst__rowSearchFocus',
              className,
            ]
              .filter(Boolean)
              .join(' ')}
            style={{
              opacity: isDraggedDescendant ? 0.5 : 1,
              ...style,
            }}
          >
            {handle}
            <div
              className={[
                'rst__rowContents',
                !canDrag && 'rst__rowContentsDragDisabled',
              ]
                .filter(Boolean)
                .join(' ')}
              onMouseEnter={() => setHover(true)}
              onMouseLeave={() => setHover(false)}
            >
              <List>
                <List.Item className="py-9">
                  {hover && (
                    <List.Content className="ml-30 pt-5" floated="right">
                      <Button.Group size="tiny">
                        {updateFile && node.isDirectory && (
                          <Button
                            onClick={() => setInputOpen(node.title)}
                            className="px-9"
                          >
                            <Icon.Group>
                              <Icon name="folder" />
                              <Icon corner name="add" />
                            </Icon.Group>
                          </Button>
                        )}
                        {updateFile && (
                          <Button
                            icon="pencil"
                            onClick={() => setRenameOpen(node)}
                          />
                        )}
                        {deleteFile && (
                          <Button
                            icon="trash alternate"
                            onClick={() => {
                              setDeleteOpen(nodeProps);
                            }}
                          />
                        )}
                      </Button.Group>
                    </List.Content>
                  )}
                  <List.Content verticalAlign="middle" className="mr-120">
                    <List.Header
                      className={isSearchMatch ? 'text-primary pt-3' : 'pt-3'}
                    >
                      {node.isDirectory ? (
                        node.title
                      ) : (
                        <Link to={`/study/${studyId}/documents/${node.kfId}`}>
                          {node.title}
                        </Link>
                      )}
                    </List.Header>
                    {node.isDirectory ? (
                      <List.Description
                        className={isSearchMatch ? 'text-primary' : 'text-grey'}
                      >
                        {node.children ? node.children.length : 'No'} items
                        included
                      </List.Description>
                    ) : (
                      <List.Description
                        className={isSearchMatch ? 'text-primary' : 'text-grey'}
                      >
                        <List horizontal className="noPadding" size="tiny">
                          {node.kfId && <List.Item>{node.kfId}</List.Item>}
                          <List.Item>{node.fileType}</List.Item>
                          <List.Item>
                            <TimeAgo
                              date={node.modified}
                              live={false}
                              title={longDate(node.modified)}
                            />
                          </List.Item>
                        </List>
                      </List.Description>
                    )}
                  </List.Content>
                </List.Item>
              </List>
            </div>
          </div>,
        )}
      </div>
    </div>
  );
};

export default withRouter(FolderItem);
