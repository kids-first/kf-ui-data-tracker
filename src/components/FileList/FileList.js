import React, {useState, Fragment} from 'react';
import PropTypes from 'prop-types';
import classes from 'classnames';
import FileElement from './FileElement';
import ReactPaginate from 'react-paginate';
import NotificationBar from '../NotificationBar/NotificationBar';
import {fileSortedVersions} from '../../common/fileUtils';
/**
 * Displays list of study files
 */
const FileList = ({className, fileList, studyId}) => {
  let fileListClass = classes('FileList', className);
  const perPage = 5;
  const [page, setPage] = useState(0);
  const handlePageClick = data => {
    setPage(data.selected);
  };
  const pageCount = () => {
    if (Math.ceil(fileList.length / perPage) < page + 1) {
      setPage(page - 1);
    }
    return Math.ceil(fileList.length / perPage);
  };
  const offset = fileList.slice(perPage * page, perPage * page + perPage);

  const showNotification = () => {
    var statusList = [];
    var i;
    for (i in fileList) {
      const sortedVersions = fileSortedVersions(fileList[i].node);
      const versionStatus =
        sortedVersions.length > 0 ? sortedVersions[0].node.state : null;
      statusList.push(versionStatus);
    }
    if (statusList.includes('CHN')) {
      return (
        <NotificationBar
          borderColor="rgba(221,31,42,1)"
          backgroundColor="rgba(221,31,42,0.1)"
          icon="error"
          message="Changes Needed"
          note="The DRC has requested that you make changes to some of your documents."
        />
      );
    } else {
      return null;
    }
  };

  return (
    <Fragment>
      {showNotification()}
      <ul className={fileListClass}>
        {fileList.length ? (
          offset.map(({node}) => (
            <FileElement key={node.kfId} fileListId={studyId} fileNode={node} />
          ))
        ) : (
          <h3 className="FileList--Empty">You don't have any files yet.</h3>
        )}
      </ul>
      {pageCount() > 0 && (
        <Fragment>
          <p className="m-0 text-sm text-mediumGrey text-right">
            {offset.length} of {fileList.length} files
          </p>
          <ReactPaginate
            previousLabel={'<'}
            nextLabel={'>'}
            breakLabel={'...'}
            breakClassName={'break-me'}
            pageCount={pageCount()}
            marginPagesDisplayed={2}
            pageRangeDisplayed={3}
            forcePage={page}
            onPageChange={handlePageClick}
            containerClassName={'Pagination'}
            subContainerClassName={'Pagination'}
            activeClassName={'Pagination--active'}
          />
        </Fragment>
      )}
    </Fragment>
  );
};

FileList.propTypes = {
  /** Any additional classes to be applied to the study list*/
  className: PropTypes.string,
  /** Array of study object*/
  fileList: PropTypes.array,
  /** Kids First unique study identifier (SD_XXXXXXXX) */
  studyId: PropTypes.string.isRequired,
};

FileList.defaultProps = {
  className: null,
  fileList: [],
  studyId: null,
};

export default FileList;
