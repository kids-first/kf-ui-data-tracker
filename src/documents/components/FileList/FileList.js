import React, {useState, Fragment} from 'react';
import PropTypes from 'prop-types';
import FileElement from './FileElement';
import {fileLatestStatus} from '../../utilities';
import {
  Header,
  Icon,
  Segment,
  Message,
  Pagination,
  Table,
  Checkbox,
} from 'semantic-ui-react';

/**
 * Displays list of study files
 */
const FileList = ({
  fileList,
  studyId,
  isAdmin,
  updateFile,
  updateError,
  downloadFileMutation,
  deleteFile,
  selection,
  setSelection,
}) => {
  const perPage = 10;
  const [page, setPage] = useState(1);
  const handlePageClick = (e, {activePage}) => {
    setPage(activePage);
  };
  const onSelectOne = fileKfID => {
    if (selection.includes(fileKfID)) {
      setSelection(selection.filter(id => id !== fileKfID));
    } else {
      setSelection([...selection, fileKfID]);
    }
  };
  const onSelectAll = () => {
    if (selection.length === fileList.length) {
      setSelection([]);
    } else {
      setSelection(fileList.map(({node}) => node.kfId));
    }
  };
  let pageCount = Math.ceil(fileList.length / perPage);
  let paginatedList = fileList.slice(
    perPage * (page - 1),
    perPage * (page - 1) + perPage,
  );
  return (
    <Fragment>
      {fileList.filter(obj => fileLatestStatus(obj.node) === 'CHN').length >
        0 && (
        <Message
          negative
          icon="warning circle"
          header="Changes Needed"
          content="The DRC has requested that you make changes to some of your documents."
        />
      )}
      {updateError && (
        <Message
          negative
          icon="warning circle"
          header="Error"
          content={updateError.message}
        />
      )}
      {fileList.length ? (
        <>
          <Table stackable selectable compact="very" celled>
            <Table.Header>
              <Table.Row>
                <Table.HeaderCell
                  textAlign="center"
                  width="1"
                  onClick={e => {
                    e.stopPropagation();
                    onSelectAll();
                  }}
                >
                  <Checkbox
                    data-testid="file-select-all"
                    checked={selection.length === fileList.length}
                  />
                </Table.HeaderCell>
                <Table.HeaderCell textAlign="center" width="1">
                  Type
                </Table.HeaderCell>
                <Table.HeaderCell className="px-20">
                  Document Details
                </Table.HeaderCell>
                <Table.HeaderCell textAlign="center">Tags</Table.HeaderCell>
                <Table.HeaderCell textAlign="center">Actions</Table.HeaderCell>
              </Table.Row>
            </Table.Header>
            <Table.Body>
              {paginatedList.map(({node}) => (
                <FileElement
                  key={node.kfId}
                  fileListId={studyId}
                  fileNode={node}
                  isAdmin={isAdmin}
                  updateFile={updateFile}
                  selection={selection}
                  onSelectOne={onSelectOne}
                  deleteFile={isAdmin ? deleteFile : null}
                  downloadFileMutation={downloadFileMutation}
                />
              ))}
            </Table.Body>
          </Table>
        </>
      ) : (
        <Segment basic>
          <Header disabled icon textAlign="center">
            <Icon name="file alternate outline" />
            No documents matching your filter or search term
          </Header>
        </Segment>
      )}
      {pageCount > 1 && (
        <Segment basic textAlign="right">
          Showing {perPage} of {fileList.length} files{' '}
          <Pagination
            firstItem={null}
            lastItem={null}
            nextItem={{
              'aria-label': 'Next page',
              content: <Icon name="chevron right" />,
            }}
            prevItem={{
              'aria-label': 'Previous page',
              content: <Icon name="chevron left" />,
            }}
            activePage={page}
            totalPages={pageCount}
            onPageChange={handlePageClick}
          />
        </Segment>
      )}
    </Fragment>
  );
};

FileList.propTypes = {
  /** Array of study object*/
  fileList: PropTypes.array,
  /** Kids First unique study identifier (SD_XXXXXXXX) */
  studyId: PropTypes.string.isRequired,
};

FileList.defaultProps = {
  fileList: [],
  studyId: null,
};

export default FileList;
