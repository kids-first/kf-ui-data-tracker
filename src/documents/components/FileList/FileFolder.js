import {
  Button,
  Divider,
  Grid,
  Header,
  Icon,
  Input,
  List,
  Modal,
} from 'semantic-ui-react';
import React, {useState} from 'react';
import SortableTree, {insertNode} from 'react-sortable-tree';
import {
  generatePath,
  keyedFiles,
  listToTree,
  treeToList,
} from '../../../common/treeDataUtils';

import FolderItem from './FolderItem';
import {searchTree} from '../../../common/treeDataUtils';
import {v4 as uuidv4} from 'uuid';
import {withRouter} from 'react-router-dom';

/**
 * Displays study files in folder view
 */
const FileFolder = ({
  fileList,
  updateFile,
  deleteFile,
  match,
  history,
  allowUploadFile,
  allowUploadVersion,
  folderMode,
  setFolderMode,
  onUpload,
}) => {
  const filesFlat = keyedFiles(fileList, match);
  const filesTree = listToTree(filesFlat);
  const [searchString, setSearchString] = useState('');
  const [treeData, setTreeData] = useState(filesTree);
  const [inputOpen, setInputOpen] = useState(null);
  const [nameInput, setNameInput] = useState('');
  const [renameOpen, setRenameOpen] = useState(null);
  const [renameInput, setRenameInput] = useState('');
  const [deleteOpen, setDeleteOpen] = useState(null);

  const existNames =
    fileList.length > 0 ? fileList.map(({node}) => node.name) : [];
  const existFolders =
    filesFlat.length > 0
      ? filesFlat
          .map(f => (f.isDirectory ? f.title : null))
          .filter(t => t !== null)
      : [];

  const updateTags = (kfId, newPath) => {
    const file = fileList.find(({node}) => node.kfId === kfId);
    if (!file) return;
    const fileNode = file.node;
    const oldPath = fileNode.tags.find(t => t.includes('/')) || null;
    const tags =
      oldPath !== newPath
        ? fileNode.tags.concat(newPath).filter(t => t !== oldPath && t !== '')
        : fileNode.tags;
    updateFile({
      variables: {
        kfId: fileNode.kfId,
        fileType: fileNode.fileType,
        tags: tags,
        name: fileNode.name,
      },
    });
  };

  const updateName = (kfId, newName) => {
    const fileNode = fileList.find(({node}) => node.kfId === kfId).node;
    updateFile({
      variables: {
        kfId: fileNode.kfId,
        fileType: fileNode.fileType,
        tags: fileNode.tags,
        name: newName,
      },
    });
  };

  const handleMoveNode = props => {
    var tree = treeData;
    if (props.node.isDirectory) {
      const allChildren = treeToList(props.node);
      const newFolderPath = props.nextParentNode
        ? generatePath(props, treeData, filesFlat)
        : '';
      allChildren.forEach(f => {
        searchTree({title: 'root', children: tree}, props.node.title).path =
          newFolderPath + props.node.title + '/';
        updateTags(f.kfId, newFolderPath + props.node.title + '/');
      });
    } else {
      if (props.nextParentNode) {
        searchTree(
          {title: 'root', children: treeData},
          props.node.title,
        ).path = generatePath(props, tree, filesFlat);
        updateTags(props.node.kfId, generatePath(props, treeData, filesFlat));
      } else {
        searchTree({title: 'root', children: tree}, props.node.title).path = '';
        updateTags(props.node.kfId, '');
      }
    }
    setTreeData(tree);
  };

  const handleNewFolder = () => {
    if (inputOpen === '_root_') {
      const newRootFolder = insertNode({
        treeData: treeData,
        depth: 0,
        minimumTreeIndex: treeData.length,
        newNode: {
          key: uuidv4(),
          title: nameInput,
          isDirectory: true,
          children: [],
          expanded: false,
          parentId: '',
        },
      });
      setTreeData(newRootFolder.treeData);
    } else {
      var tree = treeData;
      var selectedNode = searchTree({title: 'root', children: tree}, inputOpen);
      selectedNode.children.push({
        key: uuidv4(),
        title: nameInput,
        isDirectory: true,
        children: [],
        expanded: false,
        parentId: selectedNode.title,
      });
      setTreeData(tree);
    }
    setNameInput('');
  };

  const handleUpdateFolder = () => {
    if (renameOpen.isDirectory) {
      if (renameOpen.children.length > 0) {
        const allChildren = treeToList(renameOpen);
        if (allChildren.length > 0) {
          allChildren.forEach(f => {
            updateTags(f.kfId, f.path.replace(renameOpen.title, renameInput));
          });
        }
      }
    } else {
      updateName(renameOpen.kfId, renameInput);
    }
    var tree = treeData;
    var selectedNode = searchTree(
      {title: 'root', children: tree},
      renameOpen.title,
    );
    selectedNode.title = renameInput;
    setTreeData(tree);
    setRenameInput('');
  };

  const handleDeleteNode = () => {
    const parentNode = deleteOpen.parentNode;
    const selectedNode = deleteOpen.node;
    const allChildren = treeToList(selectedNode);
    if (selectedNode.isDirectory) {
      if (selectedNode.children.length > 0 && allChildren.length > 0) {
        allChildren.forEach(f => {
          updateTags(f.kfId, '');
        });
      }
    } else {
      deleteFile({variables: {kfId: selectedNode.kfId}});
    }
    var tree = treeData.concat(allChildren);
    if (parentNode !== null) {
      const nodeIndex = searchTree(
        {title: 'root', children: tree},
        parentNode.title,
      ).children.indexOf(selectedNode);
      searchTree(
        {title: 'root', children: tree},
        parentNode.title,
      ).children.splice(nodeIndex, 1);
      setTreeData(tree);
    } else {
      setTreeData(tree.filter(f => f.title !== selectedNode.title));
    }
  };

  const handleViewToggle = e => {
    e.preventDefault();
    e.stopPropagation();
    setFolderMode(!folderMode);
  };

  return (
    <Grid>
      <Grid.Row>
        <Grid.Column width={4}>
          <h2>Study Documents</h2>
        </Grid.Column>
        {fileList.length > 0 && (allowUploadFile || allowUploadVersion) && (
          <Grid.Column width={12} textAlign="right">
            <Input
              icon="search"
              placeholder="Search..."
              onChange={(e, {value}) => {
                setSearchString(value);
              }}
            />
            <Button.Group floated="right" className="ml-15">
              <Button
                icon="folder"
                active={folderMode}
                data-testid="file-folder-mode"
                onClick={handleViewToggle}
              />
              <Button
                icon="list"
                active={!folderMode}
                data-testid="file-list-mode"
                onClick={handleViewToggle}
              />
            </Button.Group>
            <Button
              className="ml-15"
              compact
              primary
              floated="right"
              size="large"
              icon="cloud upload"
              labelPosition="left"
              content="Upload Document"
              as="label"
              htmlFor="file"
            />
            <input
              hidden
              multiple
              id="file"
              type="file"
              onChange={e => onUpload(e.target.files[0])}
            />
            <Button
              className="ml-15"
              onClick={() => {
                setInputOpen('_root_');
              }}
            >
              <Icon.Group className="mr-5">
                <Icon name="folder" />
                <Icon corner name="add" />
              </Icon.Group>
              New folder
            </Button>
          </Grid.Column>
        )}
      </Grid.Row>
      <Grid.Row>
        <Grid.Column width={16} className="h-500-container">
          <SortableTree
            style={{width: '100%'}}
            rowHeight={60}
            searchQuery={searchString}
            treeData={treeData}
            onChange={treeData => setTreeData(treeData)}
            nodeContentRenderer={props => {
              return (
                <FolderItem
                  nodeProps={props}
                  searchString={searchString}
                  setSearchString={setSearchString}
                  updateFile={updateFile}
                  deleteFile={deleteFile}
                  studyId={match.params.kfId}
                  setInputOpen={setInputOpen}
                  setRenameOpen={setRenameOpen}
                  setDeleteOpen={setDeleteOpen}
                />
              );
            }}
            onMoveNode={handleMoveNode}
            canDrag={({node}) => !node.dragDisabled && updateFile}
            canDrop={({nextParent}) => !nextParent || nextParent.isDirectory}
          />
        </Grid.Column>
      </Grid.Row>
      <Modal open={inputOpen !== null} size="small">
        <Modal.Header>Please name your new folder</Modal.Header>
        <Modal.Content>
          <Input
            fluid
            placeholder="Folder name ..."
            value={nameInput}
            onChange={(e, {value}) => {
              setNameInput(value);
            }}
          />
          {existFolders.includes(nameInput) && (
            <small className="text-red">
              This name is used by another folder.
            </small>
          )}
        </Modal.Content>
        <Modal.Actions>
          <Button
            onClick={() => {
              setNameInput('');
              setInputOpen(null);
            }}
          >
            Cancel
          </Button>
          <Button
            primary
            disabled={
              nameInput.length === 0 || existFolders.includes(nameInput)
            }
            onClick={() => {
              setInputOpen(null);
              handleNewFolder();
            }}
          >
            Save
          </Button>
        </Modal.Actions>
      </Modal>
      <Modal open={renameOpen !== null} size="small">
        <Modal.Header>
          {renameOpen && renameOpen.isDirectory
            ? 'Please rename your folder'
            : 'Please rename your file'}
        </Modal.Header>
        <Modal.Content>
          <Input
            fluid
            placeholder={renameOpen ? renameOpen.title : 'New Name ...'}
            value={renameInput}
            onChange={(e, {value}) => {
              setRenameInput(value);
            }}
            error={existNames.includes(renameInput)}
          />
          {existNames.includes(renameInput) && (
            <small className="text-red">
              This name is used by another file.
            </small>
          )}
          {existFolders.includes(renameInput) && (
            <small className="text-red">
              This name is used by another folder.
            </small>
          )}
        </Modal.Content>
        <Modal.Actions>
          <Button
            onClick={() => {
              setRenameInput('');
              setRenameOpen(null);
            }}
          >
            Cancel
          </Button>
          <Button
            primary
            disabled={
              renameInput.length === 0 ||
              existNames.includes(renameInput) ||
              existFolders.includes(renameInput)
            }
            onClick={() => {
              setRenameOpen(null);
              handleUpdateFolder();
            }}
          >
            Save
          </Button>
        </Modal.Actions>
      </Modal>
      <Modal open={deleteOpen !== null} size="small">
        <Modal.Header>
          Please confirm the deletion of file or folder
        </Modal.Header>
        {deleteOpen && (
          <Modal.Content>
            {deleteOpen.node.isDirectory ? (
              <>
                <Header as="h4">Folder Selected:</Header>
                <List className="pl-16" size="large">
                  <List.Item>
                    <List.Icon className="mr-5" name="folder" />
                    {deleteOpen.node.title}
                  </List.Item>
                </List>
                {deleteOpen.node.children.length > 0 &&
                  treeToList(deleteOpen.node).length > 0 && (
                    <>
                      <Header as="h4">
                        Files will be reset back to root directory:
                      </Header>
                      <List className="pl-16" size="large">
                        {treeToList(deleteOpen.node).map(f => (
                          <List.Item key={f.kfId}>
                            <List.Icon className="mr-5" name="file alternate" />
                            {f.title}
                          </List.Item>
                        ))}
                      </List>
                    </>
                  )}
              </>
            ) : (
              <>
                <Header as="h4">File Selected:</Header>
                <List className="pl-16" size="large">
                  <List.Item>
                    <List.Icon className="mr-5" name="file alternate" />
                    {deleteOpen.node.title}
                  </List.Item>
                </List>
              </>
            )}
            <Divider />
            <List bulleted size="small" className="text-grey pl-16">
              <List.Item>
                File and its versions and history will be deleted.
              </List.Item>
              <List.Item>
                Folder will be and deleted with the files in it reset back to
                root direcory.
              </List.Item>
            </List>
          </Modal.Content>
        )}
        <Modal.Actions>
          <Button
            onClick={() => {
              setDeleteOpen(null);
            }}
          >
            Cancel
          </Button>
          <Button
            color="red"
            onClick={() => {
              setDeleteOpen(null);
              handleDeleteNode();
            }}
          >
            Delete
          </Button>
        </Modal.Actions>
      </Modal>
    </Grid>
  );
};

export default withRouter(FileFolder);
