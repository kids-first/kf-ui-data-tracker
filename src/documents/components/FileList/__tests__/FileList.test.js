import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import {render, cleanup} from '@testing-library/react';
import FileList from '../FileList';

afterEach(cleanup);

it('renders without files', () => {
  const tree = render(
    <MemoryRouter>
      <FileList fileList={[]} studyId={'SD_00000000'} />
    </MemoryRouter>,
  );
  expect(tree.container).toMatchSnapshot();

  const lis = tree.queryAllByTestId('file-item');
  expect(lis.length).toBe(0);

  // Hide pagination when no files exist
  const pagination = tree.container.querySelectorAll('.Pagination');
  expect(pagination.length).toBe(0);
});
