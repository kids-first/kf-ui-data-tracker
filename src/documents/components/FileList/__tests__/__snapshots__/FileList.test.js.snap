// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without files 1`] = `
<div>
  <div
    class="ui basic segment"
  >
    <div
      class="ui icon center aligned header"
    >
      <i
        aria-hidden="true"
        class="file alternate outline icon"
      />
      You don't have any documents yet.
    </div>
  </div>
</div>
`;
